### DISCUSSION TITLE
Questions about the usage of the Simulator to get best size

### DISCUSSION THREAD

**Discussion by adriacabeza** [ben-manes/caffeine#1763] [Q&A] [answered]
Hi! I recently added a logger into my workload so that I could capture the traces. In my case I decided to use the LIRS format as it was very easy to get. The reason why I'd like to use the simulator is because it allows me to choose between different maxSizes and see the hit rate. My current cache is experiencing a lot of misses and I suspect it is due thrashing so I'd like to see which would be a better optimal size. 

Given that the simulator contains a lot of policies and options, I'd like to know which would be your recommendation here. My goal is not comparing different policies to see if Caffeine (aka TinyLFU) is the best policy for my use case, I want to see which would be the best params for the cache (from what I've read a bunch of the tuneable parameters are already adapted with HillClimbing, that's awesome!). 

Which would be the `application.conf` file that would be the most similar to the current Caffeine implementation? I tried some simulations and the hit rate metrics were not similar to the ones reported by `recordStats for any of the policies (using the `resources.conf` as it is). 

Thanks in advance! 

PS: Adding as well one of the small trace file that I captured to this message in case it is helpful. 
[1.zip](https://github.com/user-attachments/files/16688393/1.zip)

**Comment by ben-manes** [comment, marked as answer]
The simulator includes caching products as well, as you probably noticed, so you can generate a chart like below. Since the hit rate is so high and the maximum size is small, I'd guess that in production you have expiration causing it to evict more aggressively? Unfortunately that is not currently supported by the simulator. It might be faster to just run your own script to calculate with, using `FakeTicker` to adjust the logical time based on the event's timestamp.

To add expiration to the simulator wouldn't be hard, just time consuming:
1. Make it the responsibility of the `PolicyActor` to track and delete, rather than each algorithm.
2. Add a removal method to `Policy`, a timestamp to the `AccessEvent`, and any relevant metrics in `PolicyStats`
3. Update all of the implementations to allow for external deletions
4. Add a trace reader that supports the timestamped log format


<details open><summary>Chart</summary>

```console
./gradlew simulator:simulate -q \
  -Dcaffeine.simulator.files.paths.0=lirs:/Users/ben/Downloads/1.zip \
  --maximumSize=10,20,30,40,50,60,70,80,90,100 \
  -Dcaffeine.simulator.policies.0=opt.Clairvoyant \
  -Dcaffeine.simulator.policies.1=product.Caffeine \
  -Dcaffeine.simulator.policies.2=product.Guava \
  --title=Experiment
```

<img src="https://github.com/user-attachments/assets/5622e39f-9f12-49f2-bf49-6071f3fba5f3" height="200">
</details>

**Reply by ben-manes** [reply]
for fun, here is a frequency histogram of your data sample.

```console
cat ~/Downloads/1.log \
  | sort \
  | uniq -c \
  | sort -hr \
  | head -n 25 \
  | awk '{printf "%-15s ", $2; for (i = 0; i < $1/10000; i++) printf "*"; print ""}'

-427125702      **************************************************************************************
439283015       **************
651699012       ****
-416970270      ****
-1466723613     ****
659027692       ****
-2124870682     ***
1328399074      ***
-431198752      ***
1573935389      ***
664201264       **
-1849809819     **
1207888672      **
-738926869      **
1827183881      **
-1985175728     **
738193143       *
2122036442      *
-1258062283     *
1033892603      *
37772915        *
1345948885      *
974446421       *
1299928157      *
1176115861      *

**Reply by ben-manes** [reply]
I fixed the chart to improve its focus on the range of interest. Caffeine is a little wobbly compared to a static W-TinyLFU, so its adaptive search has a worst case -2.5% impact, which is acceptable. I think because the size is so tiny it struggles to reach a good steady state and gets stuck in a worse configuration longer, as the step size is 6.25% of capacity for a sampling periods of 10x capacity, so e.g. its shifting by 2 units in 400 events which is too limited and noisy. Its a minor flaw where we can't get truly optimal at minuscule caches, but it hasn't been worth solving for as users just bump up to something like 512 capacity as extremely small caches don't often make a ton of sense anyway.

![hit_rate](https://github.com/user-attachments/assets/9eb0da67-b7f8-4428-ace6-8721bc1c7a09)

Anyway for your actual problem, if you use expiration then review `refreshAfterWrite`. Or if you are using weak/soft references then you are subject to the GC and if weak keys then since we use identity equality you should ensure that the canonical key is being used or you'll have extra mappings.

**Comment by ben-manes** [comment]
I did some work on turning the hill climber thanks to LLMs to assist in many rounds of experimentations over the last ~6 months. I tried various off-the-shelf strategies (such as PIDs and Extremum Seeking Controllers) without much luck. That led to trying to not simply graft other's approaches and attempting something new, while taking ideas from control theory to make it more robust. The next release's new design is described in this (exhaustingly long) [guide](https://htmlpreview.github.io/?https://github.com/ben-manes/caffeine/blob/master/wiki/adaptive-window.html).

Note that it was extremely hard to tune tiny caches like your example trace because there is very little usable history, the adjustments cause wide swings, and it prematurely settles at a poor configuration. I instead adjusted the parameters to give small caches more opportunities, which had a net positive effect but searching itself has a cost so its not universal.

At modestly sizes caches or larger (>4096) then the new design can stabilize at a better configuration. It does so slower than the existing climber and also runs validations of findings to cope with being a moving convex optimization problem (e.g. the workload shifts slowly over the day so the prior setting is no longer optimal, but perhaps too gradually to be noticed). The comparison on workload traces shows a very modest net benefit and it passes many synthetic hostile workload patterns meant to trap it at a bad set point. The trace corpus results downplay the benefits because the traces may not represent real Java application workloads and do not model TTLs, soft/weak references, distributed invalidations, workload seasonality, cache warming, and numerous other behaviors that happen in production environments. We don't know if the existing climber gets caught in a failure mode for some users, but we can determine that the new climber does not and can hope that helps if its a problem.

I suspect that for most users the redesign will go unnoticed and have no impact, which means we already did a great job and will continue to, so that is my ideal outcome.
